src/util/taproot.rs

```diff
@@ -82,6 +85,29 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
     doc="Taproot-tagged hash for the taproot signature hash", true
 );
 
+/// A leaf of tapscript
+pub enum TapLeaf {
+    /// Known script
+    Script(Script),
+    /// Uknown script hidden behind tapleaf hash
+    Hash(TapLeafHash),
+}
+
+/// Tapscript source, which may be only partially known
+pub struct TapScript(Vec<TapLeaf>);
+
+/// Internal taproot public key with tapscript data
+pub struct ScriptedPubkey {
```

Comment:
nit: 

```suggestion
pub struct ScriptPubkey {
```

Reply:
It is neither `Taproot` or `Tap` pubkey since this data type only covers variant of pubkey which has an associated script and does not cover self-tweaked single pubkey which can be used in taproot. I am also not happy with `ScriptedPubkey` name, but we need to somehow highlight that this is one of two options for source data in final taproot generation process from the internal key.

Reply:
A little more bikeshedding: `TapScriptPubkey`? (since it would only be used in the script path)

Fine with whatever name, after more thought: `ScriptPubkey` might be confused with `scriptPubkey` from BIPs